### PR TITLE
Custom dataset mapper

### PR DESCRIPTION
**Pull request by sei-hmoore**
### UPDATE 05/11/2022
Removing the custom dataset mapper plugin, adding custom `PANTrainer(Detectron2Trainer)`. This change made the implementation more simplified(reduced the amount of code needed) but less pluggable.
**Juneberry**
`juneberry/detectron2/trainer.py`
Modularization of the setup methods to make it easier for us to override in our custom PANTrainer

-------------------------------

### OUTDATED 05/02/2022
**Juneberry**
`junberry/config/model.py`
Update Detectron2 class to accept a new configuration option `dataset_mapper: Plugin`. This will allow a user to target a specific Dataset Mapper class in the model `config.json`.

`junberry/schemas/model_schema.py`
Update the schema to match the new `model.py` updates for Detectron2 class updates

`juneberry/detectron2/trainer.py`
Update dataset_mapper_setup to have a conditional to check our `config.json` for a `dataset_mapper` configuration. If we have this field populated we will use the Dataset Mapper class defined in the config. Otherwise, we will just use our default Detectron2 dataset mapper.

`juneberry/mappers/utils.py`
Utils file to hold helper functions specific to custom dataset mappers

### CODE REVIEW

File: juneberry/detectron2/data.py [hunk 1]
```diff
@@ -49,7 +49,7 @@
 from pathlib import Path
 import sys
 
-from detectron2.data import DatasetMapper
+from cornelius.detectron2.dataset_mapper import DatasetMapper
```
Comment (sei-amellinger) [inline juneberry/detectron2/data.py:52]: Is this correct?

File: juneberry/detectron2/trainer.py [hunk 2]
```diff
@@ -3,22 +3,19 @@
...
-# [DISTRIBUTION STATEMENT A] This material has been approved for public release and unlimited distribution.  Please see
-# Copyright notice for non-US Government use and distribution.
+# [DISTRIBUTION STATEMENT A] This material has been approved for public release and unlimited distribution.  Please see Copyright notice for non-US Government use and distribution.
+#
+# Carnegie Mellon® is registered in the U.S. Patent and Trademark Office by Carnegie Mellon University.
 #
 # This Software includes and/or makes use of Third-Party Software subject to its own license.
 #
-# DM21-0884
+# DM22-0044
```
Comment (sei-amellinger) [inline juneberry/detectron2/trainer.py:18]: Who instructed you to use these document markings?

File: juneberry/detectron2/trainer.py [hunk 3]
```diff
@@ -459,6 +337,139 @@ def train_distributed(self, num_gpus) -> None:
...
+        # At minimum, the iterations must be GREATER than the last step
+        if cfg.SOLVER.MAX_ITER <= cfg.SOLVER.STEPS[1] + 1:
+            logger.info("Adjusting MAX_ITER to be greater than the last solver step.")
+            cfg.SOLVER.MAX_ITER = cfg.SOLVER.STEPS[1] + 2
+        return cfg
+
+    def model_setup(self, cfg, args):
+        # This finalizes the freezes, finalizes config and sets other arguments
+        # TODO: What all can we pass in as args to default?
+        self.cfg = cfg
```
Comment (sei-amellinger) [inline juneberry/detectron2/trainer.py:447]: I think from a subalcassing perspective, shouldn't we have these three lines in the out function?

**Comment by sei-amellinger** [comment]
detetron2.py should also have new copyright.

**Review by sei-mchurilla — CHANGES_REQUESTED** [review]
Andrews comment on line 52 is run breaking

**Review by sei-mchurilla — APPROVED** [review]
Briefly reviewed code and it seems acceptable.  Tested and the mapper does not break any operational models.

File: juneberry/detectron2/data.py (hunk 1 above)
Comment (sei-hmoore) [reply juneberry/detectron2/data.py]: Reverted back to detectron2 DataMapper in https://github.com/cmu-sei/juneberry/pull/81/commits/a75354cc6dc8caa6538b3d29c3c3c557ff7a80cb

File: juneberry/detectron2/trainer.py (hunk 2 above)
Comment (sei-hmoore) [reply juneberry/detectron2/trainer.py]: Fixed in https://github.com/cmu-sei/juneberry/pull/81/commits/a75354cc6dc8caa6538b3d29c3c3c557ff7a80cb

File: juneberry/detectron2/trainer.py (hunk 3 above)
Comment (sei-hmoore) [reply juneberry/detectron2/trainer.py]: Makes sense, separated out the comments and `self.cfg = cfg` and `cfg.freeze()` in https://github.com/cmu-sei/juneberry/pull/81/commits/a75354cc6dc8caa6538b3d29c3c3c557ff7a80cb

File: juneberry/schemas/model_schema.json [hunk 4]
```diff
@@ -18,6 +18,7 @@
         "detectron2": {
             "type": "object",
             "properties": {
+                "dataset_mapper": { "$ref": "#/definitions/plugin_item" },
```
Comment (sei-nwinski) [inline juneberry/schemas/model_schema.json:21]: You'll also want to document this field in `docs/specs/model_configuration_specification.md`

File: juneberry/mappers/utils.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+# DM21-0884
+#
+# ======================================================================================================================
+
+import logging
+
+import juneberry.loader as jb_loader
+
+logger = logging.getLogger(__name__)
+
```
Comment (sei-nwinski) [inline juneberry/mappers/utils.py:30]: Is your IDE configured to follow PEP 8 style conventions? We usually stick to that and it's calling for an extra blank line in this location.

File: juneberry/detectron2/data.py [hunk 6]
```diff
@@ -155,15 +134,26 @@ def inverse(self) -> "Transform":
         pass
 
 
-def create_mapper(cfg, transforms, is_train: bool) -> DatasetMapper:
+def create_mapper(cfg, dataset_mapper, transforms, is_train: bool) -> DatasetMapper:
```
Comment (sei-nwinski) [inline juneberry/detectron2/data.py:137]: If you add an arg to an existing function, it's a good idea to confirm that all existing usages of the function are also up-to-date (Pycharm: Right-Click on function name -> Find Usages). You addressed mapper lines in the dt2 trainer, but there are also some in the dt2 evaluator. 

Running the system test leads to the following:
```
2022-05-09 11:22:51,911 ERROR create_mapper() missing 1 required positional argument: 'is_train'
Traceback (most recent call last):
  File "/juneberry/juneberry/scripting.py", line 45, in run_main
    main_fn()
  File "/juneberry/bin/jb_evaluate", line 143, in main
    evaluator.perform_evaluation()
  File "/juneberry/juneberry/evaluation/evaluator.py", line 244, in perform_evaluation
    self.evaluate_data()
  File "/juneberry/juneberry/detectron2/evaluator.py", line 186, in evaluate_data
    mapper = dt2_data.create_mapper(self.cfg, self.model_config.evaluation_transforms, False)
TypeError: create_mapper() missing 1 required positional argument: 'is_train'
TaskError - taskid:2
Command error: '['jb_evaluate', 'text_detect/dt2/ut', 'data_sets/text_detect_val.json']' returned 255
```

File: juneberry/detectron2/data.py [hunk 7]
```diff
@@ -155,15 +134,26 @@ def inverse(self) -> "Transform":
...
+    :param dataset_mapper: A string for the custom dataset mapper to use
     :param transforms: A list of transform objects
     :param is_train: True if training, false for testing
     :return: A new dataset mapper from the config with the augmentations.
     """
-    args = DatasetMapper.from_config(cfg, is_train)
+
+    # Conditional to test and load a custom dataset mapper if it has been set in our cfg
+    if dataset_mapper is not None:
+        logger.info("Using custom dataset mapper to override Detectron2 dataset mapper")
```
Comment (sei-nwinski) [inline juneberry/detectron2/data.py:149]: Are you familiar with f-strings in Python? They quickly became one of my favorite things and they're really helpful for log messages like this. Consider something along the lines of 
`logger.info(f"Overriding Detectron2 dataset mapper with the following custom dataset mapper: {dataset_mapper}")`

File: juneberry/detectron2/data.py [hunk 8]
```diff
@@ -155,15 +134,26 @@ def inverse(self) -> "Transform":
         pass
 
 
-def create_mapper(cfg, transforms, is_train: bool) -> DatasetMapper:
+def create_mapper(cfg, dataset_mapper, transforms, is_train: bool) -> DatasetMapper:
     """
     Creates a dataset mapper from the config including the provided transforms.
     :param cfg: The detectron2 config object
+    :param dataset_mapper: A string for the custom dataset mapper to use
```
Comment (sei-nwinski) [inline juneberry/detectron2/data.py:141]: Is this really a string if it has `.fqcn` and `.kwargs` attributes? What would the result be if you did a `type(dataset_mapper)` at this point in the code?

File: juneberry/detectron2/data.py [hunk 9]
```diff
@@ -175,4 +165,11 @@ def create_mapper(cfg, transforms, is_train: bool) -> DatasetMapper:
                 adapter.check_for_methods()
                 aug_list.append(adapter)
 
-    return DatasetMapper(**args)
+    if dataset_mapper is not None:
```
Comment (sei-nwinski) [inline juneberry/detectron2/data.py:168]: This bit confused me. It made ask "What was the point of constructing the mapper in L150 if you were just going to construct it again with a different set of args and return the second one?

Would it make sense to do all the work to build the aug_list and then worry about any data_mapper steps?

**Comment by sei-nwinski** [comment]
I think I was able to figure out why `juneberry/trainer.py` is showing up in this PR. I believe the permissions on the file changed from 644 -> 755. Was there a reason for that? If not I think if you were to restore the original permissions then the file would disappear from this PR.

File: juneberry/mappers/utils.py (hunk 5 above)
Comment (sei-hmoore) [reply juneberry/mappers/utils.py]: File has been removed due to adjustments in implementation.

File: juneberry/schemas/model_schema.json (hunk 4 above)
Comment (sei-hmoore) [reply juneberry/schemas/model_schema.json]: Line has been removed due to adjustments in implementation.

File: juneberry/detectron2/data.py (hunk 6 above)
Comment (sei-hmoore) [reply juneberry/detectron2/data.py]: Reverted back to original expected args with new implementation.

File: juneberry/detectron2/data.py (hunk 8 above)
Comment (sei-hmoore) [reply juneberry/detectron2/data.py]: Removed comment due to changes in implementation but will make note of this for future comments.

File: juneberry/detectron2/data.py (hunk 7 above)
Comment (sei-hmoore) [reply juneberry/detectron2/data.py]: Removed log output due to changes in implementation but will make note of this for future comments.

**Comment by sei-hmoore** [comment]
> I think I was able to figure out why `juneberry/trainer.py` is showing up in this PR. I believe the permissions on the file changed from 644 -> 755. Was there a reason for that? If not I think if you were to restore the original permissions then the file would disappear from this PR.

Permissions have been changed and `juneberry/trainer.py` is no longer showing any changes

**Review by sei-nwinski — APPROVED** [review]

**Review by sei-mchurilla — APPROVED** [review]
Code functions as indented.

**Pull request merged**